src/js/plugin.js

```diff
@@ -0,0 +1,436 @@
+     * - `instance`: For class-based plugins, the instance of the plugin sub-class,
+     *   but, for basic plugins, the return value of the plugin invocation.
+     *
+     * @event pluginsetup
+     * @memberof Player
+     * @instance
+     */
+    player.trigger('pluginsetup', this.getEventHash_());
+  }
+
+  /**
+   * Each event triggered by plugins includes a hash of additional data with
+   * conventional properties.
+   *
+   * This returns that object or mutates an existing hash.
+   *
+   * @private
+   * @param  {Object} [hash={}]
+   *         An object to be used as event an event hash.
+   *
+   * @return {Object}
+   *         The event hash object with, at least, the following properties:
+   *
+   *         - `instance`: The plugin instance on which the event is fired.
+   *         - `name`: The name of the plugin.
+   *         - `plugin`: The plugin class/constructor.
+   */
+  getEventHash_(hash = {}) {
+    hash.name = this.name;
+    hash.plugin = this.constructor;
+    hash.instance = this;
+    return hash;
+  }
+
+  /**
+   * Triggers an event on the plugin object.
+   *
+   * @param  {Event|Object|string} event
+   *         A string (the type) or an event object with a type attribute.
+   *
+   * @param  {Object} [hash={}]
+   *         Additional data hash to pass along with the event. For plugins,
+   *         several properties are added to the hash:
+   *
+   *         - `instance`: The plugin instance on which the event is fired.
+   *         - `name`: The name of the plugin.
+   *         - `plugin`: The plugin class/constructor.
+   *
+   * @return {boolean}
+   *         Whether or not default was prevented.
+   */
+  trigger(event, hash = {}) {
+    return Events.trigger(this.eventBusEl_, event, this.getEventHash_(hash));
+  }
+
+  /**
+   * Handles "statechanged" events on the plugin. No-op by default, override by
+   * subclassing.
+   *
+   * @param {Event} e
+   *        An event object provided by a "statechanged" event.
+   *
+   * @param {Object} e.changes
+   *        An object describing changes that occurred with the "statechanged"
+   *        event.
+   */
+  handleStateChanged(e) {}
+
+  /**
+   * Disposes a plugin.
+   *
+   * Subclasses can override this if they want, but for the sake of safety,
+   * it's probably best to subscribe the "dispose" event.
+   */
+  dispose() {
+    const {name, player} = this;
+
+    /**
+     * Signals that a class-based plugin is about to be disposed.
+     *
+     * @event dispose
+     * @memberof Plugin
+     * @instance
+     */
+    this.trigger('dispose');
+    this.off();
+
+    // Eliminate any possible sources of leaking memory by clearing up
+    // references between the player and the plugin instance and nulling out
+    // the plugin's state and replacing methods with a function that throws.
+    player[PLUGIN_CACHE_KEY][name] = false;
+    this.player = this.state = null;
+
+    this.dispose = () => {
```

Comment:
Given more thought, I think you're right on this. What made me convince myself was realizing that this could complicate plugins that want to override `dispose` and call `super.dispose()`.